plugins/PagePerformance/Tracker/PerformanceDataProcessor.php

```diff
@@ -0,0 +1,85 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+namespace Piwik\Plugins\PagePerformance\Tracker;
+
+use Piwik\Common;
+use Piwik\Log;
+use Piwik\Plugin\Dimension\ActionDimension;
+use Piwik\Plugins\PagePerformance\Columns\TimeDomCompletion;
+use Piwik\Plugins\PagePerformance\Columns\TimeDomProcessing;
+use Piwik\Plugins\PagePerformance\Columns\TimeLatency;
+use Piwik\Plugins\PagePerformance\Columns\TimeOnLoad;
+use Piwik\Plugins\PagePerformance\Columns\TimeTransfer;
+use Piwik\Tracker;
+use Piwik\Tracker\Action;
+use Piwik\Tracker\ActionPageview;
+use Piwik\Tracker\Request;
+use Piwik\Tracker\RequestProcessor;
+use Piwik\Tracker\Visit\VisitProperties;
+
+/**
+ * Handles tracker requests containing performance data.
+ */
+class PerformanceDataProcessor extends RequestProcessor
+{
+    public function recordLogs(VisitProperties $visitProperties, Request $request)
+    {
+        /** @var null|Action $action */
+        $action = $request->getMetadata('Actions', 'action');
+
+        // update performance metrics only for non pageview requests
+        if ($action && $action instanceof ActionPageview) {
+            return;
+        }
+
+        $pageViewId = $request->getParam('pv_id');
+        $visitorId = $request->getVisitorId();
+
+        // ignore requests that can't be associated with an existing page view of a visitor
+        if (empty($pageViewId) || empty($visitorId)) {
+            return;
+        }
+
+        /** @var ActionDimension[] $performanceDimensions */
+        $performanceDimensions = [
+            new TimeLatency(),
+            new TimeTransfer(),
+            new TimeDomProcessing(),
+            new TimeDomCompletion(),
+            new TimeOnLoad()
+        ];
+
+        $valuesToUpdate = [];
+
+        foreach ($performanceDimensions as $performanceDimension) {
+            $paramValue = $request->getParam($performanceDimension->getRequestParam());
+            if ($paramValue > -1) {
+                $valuesToUpdate[] = sprintf(
+                    '%s = %s',
+                    $performanceDimension->getColumnName(),
+                    $paramValue
+                );
+            }
+        }
+
+        if (empty($valuesToUpdate)) {
+            return; // no values to update given with the request
+        }
+
+        $query = sprintf(
+            'UPDATE %1$s LEFT JOIN %2$s ON idaction_url_ref = idaction SET %3$s WHERE idvisitor = ? AND idpageview = ? AND %2$s.type = 1',
```

Comment:
Would using `idvisit` be much faster than `idvisitor`? It would otherwise need to look at many more rows? It probably also be more accurate cause should someone have multiple visits then a duplicate of idpageview becomes more likely.

Not sure I understand why `idaction_url_ref ` is used? I think I would have expected the join to be `(la.idaction = lva.idaction_url) or (lva.idaction_url = 0 and la.idaction = lva.idaction_name)`. This way it also works eg for mobile apps or in general should only action_name be given. Also I don't think the url_ref is needed but the actual ref?

Also wondering if it be better to have a select query, and then an update query `where idlink_va = X` to prevent locking issues. Cause this update statement would potentially lock quite a few rows in log_action and log_link_visit_action and should there be more tracking requests coming in at the same time, this can cause issues. AFAIK Mysql will lock all rows it needs to look at.

Reply:
For the update itself you could then also use `Tracker/Model::updateAction`

Reply:
Makes sense. Will update it.